### DISCUSSION TITLE
Dynamic Table Name With Extended Entity

### DISCUSSION THREAD

**Discussion by brendano257** [SeaQL/sea-orm#1616] [Q&A] [answered]
Hi all, I'm very new to SeaOrm and relatively new to Rust as a disclaimer. 

I'm looking to make a single entity which can have a different table name depending on attributes of the entity (the schemas are the same). I have many tables with a structured way of being named, dependent on some data (each have 10's of millions of rows, which is why they're separated by table). 

My thought was that the Entity will store these two attributes, and use them to construct the table name in the impl for EntityName, but I'm running into lifetime issues. Is this even possible, or should I be looking at ways of storing a lookup of pre-generated entities instead? The below is the (non-working) gist of what I'm saying, along with the error.

```
#[derive(Copy, Clone, Default, Debug, DeriveEntity)]
pub struct Entity<'a> {
    part_one: &'a str,
    part_two: &'a str,
}

impl <'a> EntityName for Entity<'a> {
    fn table_name(self) -> &'a str {
        &format!("thing_{}_{}", self.part_one, self.part_two)
    }
}
```

This produces: 
```
   |
28 | pub struct Entity<'a> {
   |            ^^^^^^ expected lifetime parameters
   |
   = note: assuming a `'static` lifetime...
help: indicate the anonymous lifetimes
   |
28 | pub struct Entity<'_><'a> {
   |                  ++++++++
```

Which is fairly baffling to me, because I _think_ I already provided the necessary lifetime. What am I missing here?

Thanks,
Brendan

**Comment by tyt2y3** [comment, marked as answer]
You can't put a less than static lifetime on Entity.

I would suggest you to use `ArrayString`, which implements `Copy`, or use lazy_static to allocate those strings.
There were other threads about making a custom Entity like this, but I did not know how it goes and whether there are other blockers.

**Reply by brendano257** [reply]
Thanks @tyt2y3, I was able to find a couple threads with better keywords for searching. ArrayString looks like you still run into the problem of needing to return an `&str` that isn't owned by the table_name scope, but lazy_static looks promising. I'll give that a shot. If I'm not mistaken, macro-based solutions won't work for me because I specifically *don't* want these to be different types, e.g. regardless of the table they come from, I'd like to be able to have `Vec<ModelName>`, etc.

**Comment by billy1624** [comment]
Hey @brendano257, I just make a demo https://github.com/SeaQL/sea-orm/pull/1622

**Reply by brendano257** [reply]
Hi Billy, thanks so much for the clear example and making it a PR, too! I've added a comment to it about whether doing this looses the ability to use the ORM methods like `Entity::find_by_id`, etc.
